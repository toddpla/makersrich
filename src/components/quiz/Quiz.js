import React from 'react'
import { connect } from 'react-redux'
import Question from './Question'
import Answer from './Answer'
import { startSendResult, startGetQuestion, clearQuiz } from '../../actions/quiz'
import { startUpdatePlayer, startDebitPlayer, updatePlayer } from '../../actions/auth'
import { QUESTION_PRICE } from '../../constants'

export class Quiz extends React.Component {

  constructor(props) {
    super(props);
    console.log(props);

    if ( this.canAffordQuestion() ) { this.props.startGetQuestion() }

    this.state = {
      questionCount: 0,
    }
  }

  canAffordQuestion = () => {
    return this.props.auth.cash >= QUESTION_PRICE
  }

  sendResultToFirebase = (submission) => {
    if (submission.result === true) {
      this.props.auth.sessionQuestions.push(submission)
      this.props.startSendResult({...submission})
      this.setState({
        correctQuestions: this.state.correctQuestions + 1
      })
      }
    }

  removeQuiz = () => {
    var quizContainer = document.getElementById('quiz-container')
    var quiz = document.getElementById('quiz')
    quizContainer.removeChild(quiz)
  }

  handleClick = (answerIndex) => {
    this.props.startDebitPlayer(QUESTION_PRICE).then(() => {
      const submission = {
        uid: this.props.auth.uid,
        questionId: this.props.quiz.id,
        result: answerIndex.toString() === this.props.quiz.correctAnswer
      }
      this.sendResultToFirebase(submission)

      this.props.clearQuiz()
      if (this.canAffordQuestion()) {
        this.props.startGetQuestion(submission.uid)
      }
      this.setState({
        questionCount: this.state.questionCount + 1
      })
      if (this.state.questionCount === 5) {
        alert('you did it fam!')
      }
    })
  }

  render() {

    return(
      <div id='quiz-container'>
        <h1>Welcome to the quiz!</h1>
          { (Object.keys(this.props.quiz).length !== 0 && this.props.quiz.constructor === Object) ? (
            <div id='quiz'>
              <Question question={this.props.quiz.question}/>
              {this.props.quiz.answers.map((answer, i) => <Answer key={i} id={i} answer={answer} handleClick={this.handleClick} />)}
            </div>
          )
            :
          (
            <div>
              <h2>Come back when you have more money</h2>
            </div>
          )}
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
    quiz: state.quiz,
    auth: state.auth,
    questions: state.questions
})

const mapDispatchToProps = dispatch => ({
  startSendResult: (result) => dispatch(startSendResult(result)),
  startGetQuestion: () => dispatch(startGetQuestion()),
  startUpdatePlayer: (updates) => dispatch(startUpdatePlayer(updates)),
  startDebitPlayer: (amount) => dispatch(startDebitPlayer(amount)),
  clearQuiz: () => dispatch(clearQuiz()),
  updatePlayer: (updates) => dispatch(updatePlayer(updates))
})

export default connect(mapStateToProps, mapDispatchToProps)(Quiz)
