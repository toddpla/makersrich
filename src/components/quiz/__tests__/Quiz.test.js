import React from 'react';
import { shallow } from 'enzyme'
import { Quiz } from '../Quiz'

let wrapper, startSendResult, startGetQuestion, quiz

beforeEach(function() {
  startGetQuestion = jest.fn()

  quiz = {
    id: 8,
    answers: [
      'one',
      'two',
      'three',
      'four'
    ],
    correctAnswer: 'two',
    question: 'What comes after one?'
  }
  wrapper = shallow(
    <Quiz
      quiz={quiz}
      startGetQuestion={startGetQuestion}
    />
  )
});

test('it renders Quiz', () => {
  expect(wrapper).toMatchSnapshot()
})
