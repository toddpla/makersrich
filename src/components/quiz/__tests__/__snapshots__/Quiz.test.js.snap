// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz it renders Quiz 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Quiz
    auth={
      Object {
        "cash": 200,
        "sessionQuestions": Array [],
        "uid": "123",
      }
    }
    clearQuiz={[MockFunction]}
    quiz={
      Object {
        "answers": Array [
          "one",
          "two",
          "three",
          "four",
        ],
        "correctAnswer": "1",
        "id": 8,
        "question": "What comes after one?",
      }
    }
    startDebitPlayer={[MockFunction]}
    startGetQuestion={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    startSendResult={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          Welcome to the quiz!
        </h1>,
        <div
          id="quiz"
        >
          <Question
            question="What comes after one?"
          />
          <Answer
            answer="one"
            correct={false}
            handleClick={[Function]}
            id={0}
          />
          <Answer
            answer="two"
            correct={true}
            handleClick={[Function]}
            id={1}
          />
          <Answer
            answer="three"
            correct={false}
            handleClick={[Function]}
            id={2}
          />
          <Answer
            answer="four"
            correct={false}
            handleClick={[Function]}
            id={3}
          />
        </div>,
      ],
      "id": "quiz-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Welcome to the quiz!",
        },
        "ref": null,
        "rendered": "Welcome to the quiz!",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Question
              question="What comes after one?"
            />,
            Array [
              <Answer
                answer="one"
                correct={false}
                handleClick={[Function]}
                id={0}
              />,
              <Answer
                answer="two"
                correct={true}
                handleClick={[Function]}
                id={1}
              />,
              <Answer
                answer="three"
                correct={false}
                handleClick={[Function]}
                id={2}
              />,
              <Answer
                answer="four"
                correct={false}
                handleClick={[Function]}
                id={3}
              />,
            ],
          ],
          "id": "quiz",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "question": "What comes after one?",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "class",
            "props": Object {
              "answer": "one",
              "correct": false,
              "handleClick": [Function],
              "id": 0,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "class",
            "props": Object {
              "answer": "two",
              "correct": true,
              "handleClick": [Function],
              "id": 1,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "class",
            "props": Object {
              "answer": "three",
              "correct": false,
              "handleClick": [Function],
              "id": 2,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "3",
            "nodeType": "class",
            "props": Object {
              "answer": "four",
              "correct": false,
              "handleClick": [Function],
              "id": 3,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Welcome to the quiz!
          </h1>,
          <div
            id="quiz"
          >
            <Question
              question="What comes after one?"
            />
            <Answer
              answer="one"
              correct={false}
              handleClick={[Function]}
              id={0}
            />
            <Answer
              answer="two"
              correct={true}
              handleClick={[Function]}
              id={1}
            />
            <Answer
              answer="three"
              correct={false}
              handleClick={[Function]}
              id={2}
            />
            <Answer
              answer="four"
              correct={false}
              handleClick={[Function]}
              id={3}
            />
          </div>,
        ],
        "id": "quiz-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Welcome to the quiz!",
          },
          "ref": null,
          "rendered": "Welcome to the quiz!",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Question
                question="What comes after one?"
              />,
              Array [
                <Answer
                  answer="one"
                  correct={false}
                  handleClick={[Function]}
                  id={0}
                />,
                <Answer
                  answer="two"
                  correct={true}
                  handleClick={[Function]}
                  id={1}
                />,
                <Answer
                  answer="three"
                  correct={false}
                  handleClick={[Function]}
                  id={2}
                />,
                <Answer
                  answer="four"
                  correct={false}
                  handleClick={[Function]}
                  id={3}
                />,
              ],
            ],
            "id": "quiz",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "question": "What comes after one?",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "class",
              "props": Object {
                "answer": "one",
                "correct": false,
                "handleClick": [Function],
                "id": 0,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "class",
              "props": Object {
                "answer": "two",
                "correct": true,
                "handleClick": [Function],
                "id": 1,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "class",
              "props": Object {
                "answer": "three",
                "correct": false,
                "handleClick": [Function],
                "id": 2,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "3",
              "nodeType": "class",
              "props": Object {
                "answer": "four",
                "correct": false,
                "handleClick": [Function],
                "id": 3,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
